### DISCUSSION TITLE
Issue getting tabout working

### DISCUSSION THREAD

**Discussion by shamindras** [abecodes/tabout.nvim#42] [Q&A] [answered]
Thank you for creating this package.

I've tried to follow your [installation instructions](https://github.com/abecodes/tabout.nvim/tree/c37ce26f316a2df693140450b8def41e89c0f57e#-installation) carefully in [my dotfiles](https://github.com/shamindras/mydots/tree/1f72fb1ad022fd68f2deab1673facfac0023f497/config/nvim). I'm having the same issue as [here](https://github.com/abecodes/tabout.nvim/issues/41), i.e., tabout only indents inside matching pairs of parenthesis/brackets/braces, but does not tab out of them.

First I added the following lines to my general [packer installation](https://github.com/shamindras/mydots/blob/1f72fb1ad022fd68f2deab1673facfac0023f497/config/nvim/lua/shamindras/plugins-setup.lua#L120-L125):

```lua
  -- Lua
  use({
    'abecodes/tabout.nvim',
    wants = { 'nvim-treesitter' }, -- or require if not used so far
    after = { 'nvim-cmp' }, -- if a completion plugin is using tabs load it before
  })
```

Next I did the setup as follows in a [separate file](https://github.com/shamindras/mydots/blob/1f72fb1ad022fd68f2deab1673facfac0023f497/config/nvim/lua/shamindras/plugins/tabout.lua) named `tabout.lua`:

```lua
-- import tabout plugin safely
local tabout_status, tabout = pcall(require, 'tabout')
if not tabout_status then
  return
end

tabout.setup({
  tabkey = '<Tab>', -- key to trigger tabout, set to an empty string to disable
  backwards_tabkey = '<S-Tab>', -- key to trigger backwards tabout, set to an empty string to disable
  act_as_tab = true, -- shift content if tab out is not possible
  act_as_shift_tab = false, -- reverse shift content if tab out is not possible (if your keyboard/terminal supports <S-Tab>)
  default_tab = '<C-t>', -- shift default action (only at the beginning of a line, otherwise <TAB> is used)
  default_shift_tab = '<C-d>', -- reverse shift default action,
  enable_backwards = true, -- well ...
  completion = true, -- if the tabkey is used in a completion pum
  tabouts = {
    { open = "'", close = "'" },
    { open = '"', close = '"' },
    { open = '`', close = '`' },
    { open = '(', close = ')' },
    { open = '[', close = ']' },
    { open = '{', close = '}' },
  },
  ignore_beginning = true, --[[ if the cursor is at the beginning of a filled element it will rather tab out than shift the content ]]
  exclude = {}, -- tabout will ignore these filetypes
})
```

Lastly, I made sure that [this file was called](https://github.com/shamindras/mydots/blob/1f72fb1ad022fd68f2deab1673facfac0023f497/config/nvim/init.lua#L17-L19) in my init.lua

```lua
require('shamindras.plugins.treesitter')
require('shamindras.plugins.tabout')
require('shamindras.plugins.gitsigns')
```

I made sure that I did a `:PackerInstall` to ensure `tabout` was successfully installed.

**Problem**

When I restart nvim, it does so without any issue, but `tabout` does not give the intended result when pressing `<Tab>` inside a pair of matching parentheses in _insert mode_. That is, when I have `($)` (where `$` represents the insert mode cursor position after autopairs matches `(` with `)`), pressing `<Tab>` just causes indentation of width 4 spaces.

What am I doing wrong, and how should I proceed to fix this?

**Comment by abecodes** [comment, marked as answer]
Hi @shamindras 

Thanks for the detailed explanation 😃 

Two scenarios are coming to mind for this kind of problem:

- tabout is not installed correctly
- wrong syntax

Lets start with the first one. If you enter command mode and type `:tab` and then press the <TAB> key, you should see smth similar to this:

![Screenshot 2023-02-04 at 17 47 11](https://user-images.githubusercontent.com/12407796/216779740-f99c468c-cbab-4b86-b945-8c911a7d6da8.png)

If you see this, tabout is running fine and smth is off with your syntax.

Tabout.nvim uses treesitter to determine in which node you are and where to jump to. If that does not happen, treesitter can not translate your syntax into a correct node. This may be due to an issue on treesitter's side or you are using invalid syntax (depending on your language, this may also happen in comment strings).

But not all is lost, you can try it with the branch `feature/tabout-md` which has a fallback for such cases.

```
use({
    'abecodes/tabout.nvim',
    branch = 'feature/tabout-md',
    wants = { 'nvim-treesitter' }, -- or require if not used so far
    after = { 'nvim-cmp' }, -- if a completion plugin is using tabs load it before
  })
```

Let me know if that fixes your issue. If not, please provide a code sample 😄

**Reply by shamindras** [reply]
Thank you @abecodes for creating this package, and your helpful response 🎉 :

>	- tabout is not installed correctly

I just checked, and it is working as you expected.

![CleanShot 2023-02-04 at 16 16 44@2x](https://user-images.githubusercontent.com/7627188/216790358-c1b3c6bb-39bc-4258-bc08-81b0e952c0d0.png)
ng this plugin, and for your thoughtful and helpful response.

> - wrong syntax

It is actually working! I tested it on the `tabout.lua` file as @Fau818 had done [here](https://github.com/abecodes/tabout.nvim/issues/41#issue-1561155440). It seems we had both picked autopairs `[]` which happened to not work in `lua`! I tested it on matching `()`, `{}` in the `tabout.lua` file and it works as intended. I got the idea from you per [your comment](https://github.com/abecodes/tabout.nvim/issues/41#issuecomment-1416803713) in that discussion 👍🏾 .

I will keep in mind the branch of tabout that **ignores** treesitter going forward. A quick question on this though, why not make this version an **option** in the setup, which can be **toggled** in the `require('tabout').setup({})` config, rather than installed separately using a branch in `Packer`? It seems like it would be useful to toggle quickly between the 2 approaches depending on _file-specific_ treesitter issues. Just a thought.

**Reply by abecodes** [reply]
Thanks for coming back @shamindras 

That branch was actually a fix for an issue (for a similar reason) that appears in comments or md files.
I planned to merge it months ago, but I never heard back, so I forgot about it xD
I am not sure if it needs a toggle, since it will only kick in when treesitter is failing. Let me know if you have one ;)

**Reply by shamindras** [reply]
> That branch was actually a fix for an issue (for a similar reason) that appears in comments or md files. I planned to merge it months ago, but I never heard back, so I forgot about it xD I am not sure if it needs a toggle, since it will only kick in when treesitter is failing. Let me know if you have one ;)

Thanks for the clarification @abecodes - appreciate it. And great work on this package, it is very helpful.

If you are asking me to write the toggle, let me develop my skills in `lua` more and come back to you 😄
